pysd/py_backend/output.py

```diff
@@ -0,0 +1,519 @@
+        return self.handler.postprocess(**kwargs)
+
+    def add_run_elements(self, model, run_elements):
+        """ Delegating the addition of results with run cache in the output
+        object to the appropriate handler."""
+        self.handler.add_run_elements(model, run_elements)
+
+
+class OutputHandlerInterface(metaclass=abc.ABCMeta):
+    """
+    Interface for the creation of different output type handlers.
+    """
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        return (hasattr(subclass, 'initialize') and
+                callable(subclass.initialize) and
+                hasattr(subclass, 'update') and
+                callable(subclass.update) and
+                hasattr(subclass, 'postprocess') and
+                callable(subclass.postprocess) and
+                hasattr(subclass, 'add_run_elements') and
+                callable(subclass.add_run_elements) or
+                NotImplemented)
+
+    @abc.abstractmethod
+    def initialize(self, model, capture_elements):
+        """
+        Create the results object and its elements based on capture_elemetns.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update(self, model, capture_elements):
+        """
+        Update the results object at each iteration at which resutls are
+        stored.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def postprocess(self, **kwargs):
+        """
+        Perform different tasks at the time of returning the results object.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def add_run_elements(self, model, capture_elements):
+        """
+        Add elements with run cache to the results object.
+        """
+        raise NotImplementedError
+
+
+class DatasetHandler(OutputHandlerInterface):
+    """
+    Manages simulation results stored as netCDF4 Dataset.
+    """
+
+    def __init__(self, out_file):
+        self.out_file = out_file
+        self.step = 0
+        self.ds = None
+
+    def initialize(self, model, capture_elements):
+        """
+        Creates a netCDF4 Dataset and adds model dimensions and variables
+        present in the capture elements to it.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+
+        """
+        self.ds = nc.Dataset(self.out_file, "w")
+
+        # defining global attributes
+        self.ds.description = "Results for simulation run on" \
+            f"{t.ctime(t.time())} using PySD version {__version__}"
+        self.ds.model_file = model.py_model_file
+        self.ds.timestep = f"{model.components.time_step()}"
+        self.ds.initial_time = f"{model.components.initial_time()}"
+        self.ds.final_time = f"{model.components.final_time()}"
```

Comment:
Do not use `model.components.time_step` and similar for control variables.
Better use 'model.time.time_step`...

Reply:
This implementation is only valid when final time, time step and saveper are constants, maybe it would be nice to have an option to save them with time dependence.

Reply:
> This implementation is only valid when final time, time step and saveper are constants, maybe it would be nice to have an option to save them with time dependence.

Is there a test model that I could use for inspiration on that?

Reply:
> Do not use `model.components.time_step` and similar for control variables. Better use 'model.time.time_step`...

Done

Reply:
Great thanks :)

Check the following test model, which is also used in` pytest_inegration_vensim_pathway.py`:
https://github.com/SDXorg/test-models/blob/master/tests/control_vars/test_control_vars.mdl

Note that this is also documented by Vensim:
https://www.vensim.com/documentation/timeboundsdates.html
> NOTE All of the control parameters that you set from this dialog can also be set directly by writing equations for the control parameters.  By default all except SAVEPER are constants, but they do not need to be constants.  The example model ball.mdl demonstrates the writing of an equation for TIME STEP that depends on conditions detected in the model.  If you do make changes to control parameters in this model, you will get a message that there were problems updating and that you should use the Equation Editor.  Some of the changes you make might still hold.

Reply:
The results should now work regardles of whether saveper, time_step and final_time have run or cache type.
New values after each iteration are added based on index value, not the actual time, and there's no time preallocation.
I made changes to the way the Dataset global attributes are defined, too. If the control variables have cache step, then the values of the timestep and final_time global attributes are set to "Variable".